### PR TITLE
[Feature] Add flat cosine lr updater

### PR DESCRIPTION
**Pull request by wangg12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
In some cases we want to anneal lr after a 'flat lr' period. 

## Modification
Add a flat + cosine annealing lr scheduler similar to https://github.com/fastai/fastai/blob/master/fastai/callback/schedule.py#L128.

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zhouzaida** [comment]
hi @wangg12 , please sign the CLA. Besides, unittest should be added.

**Comment by zhouzaida** [comment]
please fix the lint

File: mmcv/runner/hooks/lr_updater.py [hunk 1]
```diff
@@ -278,6 +278,46 @@ def get_lr(self, runner, base_lr):
         return annealing_cos(base_lr, target_lr, progress / max_progress)
 
 
+@HOOKS.register_module()
+class FlatCosineAnnealingLrUpdaterHook(LrUpdaterHook):
+    """Flat + Cosine lr schedule.
+
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:284]: please add docstring

File: mmcv/runner/hooks/lr_updater.py [hunk 2]
```diff
@@ -278,6 +278,46 @@ def get_lr(self, runner, base_lr):
...
+    https://github.com/fastai/fastai/blob/master/fastai/callback/schedule.py#L128
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:285]: Modified from
https://github.com/fastai/fastai/blob/master/fastai/callback/schedule.py#L128

File: mmcv/runner/hooks/lr_updater.py [hunk 3]
```diff
@@ -278,6 +278,46 @@ def get_lr(self, runner, base_lr):
...
+    """
+
+    def __init__(self,
+                 start_pct=0.75,
+                 min_lr=None,
+                 min_lr_ratio=None,
+                 **kwargs):
+        assert (min_lr is None) ^ (min_lr_ratio is None)
+        assert start_pct >= 0 and start_pct < 1
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:294]: assert start_pct >= 0 and start_pct < 1, '"start_pct" must be in range (0, 1]'

File: mmcv/runner/hooks/lr_updater.py [hunk 4]
```diff
@@ -278,6 +278,57 @@ def get_lr(self, runner, base_lr):
...
+    """Flat + Cosine lr schedule.
+
+    Modified from https://github.com/fastai/fastai/blob/master/fastai/callback/schedule.py#L128 # noqa: E501
+
+    Args:
+        start_pct (float): When to start annealing the learning rate
+            after the percentage of the total training steps.
+            The value should be in range [0, 1).
+            Default: 0.75
+        min_lr (float): The minimum lr. Default: None.
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:292]: Because of `Default: None`, `min_lr (float)` should be changed to `min_lr (float, optional)`

File: mmcv/runner/hooks/lr_updater.py [hunk 5]
```diff
@@ -278,6 +278,57 @@ def get_lr(self, runner, base_lr):
...
+        min_lr_ratio (float): The ratio of minimum lr to the base lr.
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:293]: the same as min_lr

File: tests/test_runner/test_hooks.py [hunk 6]
```diff
@@ -355,6 +355,97 @@ def test_cosine_runner_hook(multi_optimziers):
     hook.writer.add_scalars.assert_has_calls(calls, any_order=True)
 
 
+@pytest.mark.parametrize('multi_optimziers', (True, False))
+def test_flat_cosine_runner_hook(multi_optimziers):
```
Comment (zhouzaida) [inline tests/test_runner/test_hooks.py:359]: maybe we should add unittest like https://github.com/open-mmlab/mmcv/blob/bf2c9fa8d21e4af903769b3d7aca20452e4b4669/tests/test_runner/test_hooks.py#L363

**Comment by zhouzaida** [comment]
Is it necessary to add FlatCosineAnnealingMomentumUpdaterHook

**Comment by wangg12** [comment]
> Is it necessary to add FlatCosineAnnealingMomentumUpdaterHook

Sorry. I don't know.

**Comment by zhouzaida** [comment]
> > Is it necessary to add FlatCosineAnnealingMomentumUpdaterHook
> 
> Sorry. I don't know.

got it

File: tests/test_runner/test_hooks.py [hunk 7]
```diff
@@ -355,6 +356,105 @@ def test_cosine_runner_hook(multi_optimziers):
...
+    """xdoctest -m tests/test_hooks.py test_flat_cosine_runner_hook."""
+    sys.modules['pavi'] = MagicMock()
+    loader = DataLoader(torch.ones((10, 2)))
+    runner = _build_demo_runner(multi_optimziers=multi_optimziers)
+
+    with pytest.raises(AssertionError):
+        # by_epoch should be False
+        FlatCosineAnnealingLrUpdaterHook(by_epoch=True)
```
Comment (zhouzaida) [inline tests/test_runner/test_hooks.py:368]: Does the FlatCosineAnnealingLrUpdaterHook only support `by_epoch=True`

Comment (wangg12) [reply tests/test_runner/test_hooks.py:368]: Both should be supported, but in general `by_epoch=False` is recommended.

Comment (zhouzaida) [reply tests/test_runner/test_hooks.py:368]: so why `by_epoch=True` will raise `AssertionError`?

Comment (wangg12) [reply tests/test_runner/test_hooks.py:368]: I thought it was similar to `test_one_cycle_runner_hook`, so I just copied the code from it.

Comment (zhouzaida) [reply tests/test_runner/test_hooks.py:368]: In fact, `OneCycleLrUpdaterHook` only support `by_epoch=False`

https://github.com/open-mmlab/mmcv/blob/bf2c9fa8d21e4af903769b3d7aca20452e4b4669/mmcv/runner/hooks/lr_updater.py#L488-L492

Comment (wangg12) [reply tests/test_runner/test_hooks.py:368]: OK. I will remove this test.

Comment (zhouzaida) [reply tests/test_runner/test_hooks.py]: ok

File: tests/test_runner/test_hooks.py [hunk 8]
```diff
@@ -355,6 +356,101 @@ def test_cosine_runner_hook(multi_optimziers):
...
+    loader = DataLoader(torch.ones((10, 2)))
+    runner = _build_demo_runner(multi_optimziers=multi_optimziers)
+
+    with pytest.raises(ValueError):
+        # expected float between 0 and 1
+        FlatCosineAnnealingLrUpdaterHook(start_pct=-0.1)
+
+    # add momentum scheduler
+
+    hook_cfg = dict(
```
Comment (zhouzaida) [inline tests/test_runner/test_hooks.py:372]: If `FlatCosineAnnealingLrUpdaterHook` has no corresponding `Momentum Update Hook`, the hook_cfg can be removed.

Comment (wangg12) [reply tests/test_runner/test_hooks.py]: OK.

File: mmcv/runner/hooks/lr_updater.py [hunk 9]
```diff
@@ -278,6 +278,58 @@ def get_lr(self, runner, base_lr):
...
+        if start_pct < 0 or start_pct > 1 or not isinstance(start_pct, float):
+            raise ValueError('expected float between 0 and 1 start_pct, but '
+                             f'got {start_pct}')
+        self.start_pct = start_pct
+        self.min_lr = min_lr
+        self.min_lr_ratio = min_lr_ratio
+        super(FlatCosineAnnealingLrUpdaterHook, self).__init__(**kwargs)
+
+    def get_lr(self, runner, base_lr):
+        if self.by_epoch:
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:313]: should also add unittest to check `by_epoch = True`

Comment (wangg12) [reply mmcv/runner/hooks/lr_updater.py:313]: added

**Comment by ZwwWayne** [comment]
The implementation seems very similar to `CyclicLrUpdaterHook`, can `CyclicLrUpdaterHook` meet your need?

**Comment by wangg12** [comment]
I guess no.

File: mmcv/runner/hooks/lr_updater.py [hunk 10]
```diff
@@ -278,6 +278,58 @@ def get_lr(self, runner, base_lr):
...
 
 
+@HOOKS.register_module()
+class FlatCosineAnnealingLrUpdaterHook(LrUpdaterHook):
+    """Flat + Cosine lr schedule.
+
+    Modified from https://github.com/fastai/fastai/blob/master/fastai/callback/schedule.py#L128 # noqa: E501
+
+    Args:
+        start_pct (float): When to start annealing the learning rate
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/lr_updater.py:288]: 1. pct is an unfriendly abbreviation. Suggest changing a name.

**Comment by ZwwWayne** [comment]
> I guess no.

Then how about Directly use the CosineAnealing? I think you can try `CosineAnealing` with constant warmup.

File: mmcv/runner/hooks/lr_updater.py (hunk 10 above)
Comment (wangg12) [reply mmcv/runner/hooks/lr_updater.py:288]: What name is better in your opinion?

Comment (ZwwWayne) [reply mmcv/runner/hooks/lr_updater.py:288]: How about flat_ratio=0.25?

Comment (wangg12) [reply mmcv/runner/hooks/lr_updater.py:288]: Maybe fine. But it is not very true. Because if a warmup stage is used, the flat ratio is not the given value.

Comment (ZwwWayne) [reply mmcv/runner/hooks/lr_updater.py:288]: warmup will affect no matter what the name is. Do you know the full name of pct?

Comment (wangg12) [reply mmcv/runner/hooks/lr_updater.py:288]: `start_pct` full name `start_percent`, means the percentage of the whole training progress.

Comment (ZwwWayne) [reply mmcv/runner/hooks/lr_updater.py:288]: then we should use start_percent rather than start_pct as it is really hard to guess.

Comment (wangg12) [reply mmcv/runner/hooks/lr_updater.py:288]: ok

**Review by ZwwWayne — APPROVED** [review]
